crates/noirc_evaluator/src/ssa/anchor.rs

```diff
@@ -0,0 +1,263 @@
+        let op = operation.opcode();
+
+        match operation {
+            Operation::Cast(_) => unreachable!(),
+            _ => {
+                if let std::collections::hash_map::Entry::Vacant(e) = self.map.entry(op) {
+                    let mut prev_list = HashMap::new();
+                    prev_list.insert(operation.clone(), id);
+                    e.insert(prev_list);
+                } else {
+                    self.map.get_mut(&op).unwrap().insert(operation.clone(), id);
+                }
+            }
+        }
+    }
+
+    pub fn find_similar_instruction(&self, operation: &Operation) -> Option<NodeId> {
+        let op = operation.opcode();
+        if self.map.contains_key(&op) && self.map[&op].contains_key(operation) {
+            return Some(self.map[&op][operation]);
+        }
+        None
+    }
+
+    pub fn find_similar_cast(
+        &self,
+        igen: &SsaContext,
+        operator: &Operation,
+        res_type: super::node::ObjectType,
+    ) -> Option<NodeId> {
+        match operator {
+            Operation::Cast(id) => {
+                if self.cast_map.contains_key(id) {
+                    let by_type = &self.cast_map[id];
+                    if by_type.contains_key(&res_type) {
+                        let tu = by_type[&res_type];
+                        if let Some(ins) = igen.try_get_instruction(tu) {
+                            if !ins.is_deleted() {
+                                return Some(tu);
+                            }
+                        }
+                    }
+                }
+            }
+            _ => unreachable!(),
+        }
+
+        None
+    }
+
+    fn get_mem_map(&self, a: ArrayId) -> &Vec<VecDeque<(usize, NodeId)>> {
+        &self.mem_map[&a]
+    }
+
+    pub fn get_mem_all(&self, a: ArrayId) -> &VecDeque<MemItem> {
+        &self.mem_list[&a]
+    }
+
+    pub fn use_array(&mut self, a: ArrayId, len: usize) {
+        if !self.mem_list.contains_key(&a) {
+            let def: VecDeque<(usize, NodeId)> = VecDeque::new();
+            self.mem_list.entry(a).or_insert_with(VecDeque::new);
+            self.mem_map.entry(a).or_insert_with(|| vec![def; len]);
+        }
+    }
+
+    pub fn push_mem_instruction(&mut self, ctx: &SsaContext, id: NodeId) {
+        let ins = ctx.get_instruction(id);
+        let (array_id, index, is_load) = Anchor::get_mem_op(&ins.operation);
+        self.use_array(array_id, ctx.mem[array_id].len as usize);
+        let prev_list = self.mem_list.get_mut(&array_id).unwrap();
+        let len = prev_list.len();
+        if let Some(index_value) = ctx.get_as_constant(index) {
+            let mem_idx = index_value.to_u128() as usize;
+            let last_item = prev_list.front();
+            let mut new_item = None;
+            match last_item {
+                Some(MemItem::Const(pos)) => {
+                    self.mem_map.get_mut(&array_id).unwrap()[mem_idx].push_front((*pos, id))
```

Comment:
This line seems repeated a lot, maybe we should have a helper for it.

Edit: alternatively, since all match branches do it, it would be less repetitive to return only the pos/len that actually differs from each branch and move the push_front after the match.

Reply:
Done